status-network-contracts/docs/rln.md

```diff
@@ -0,0 +1,157 @@
+## Commit-Reveal Slashing Mechanism
+
+The RLN contract implements a two-step commit-reveal scheme to prevent front-running attacks and ensure orderly
+processing of slashing operations.
+
+### Why Commit-Reveal?
+
+Without a commit-reveal scheme, a slasher could observe an on-chain transaction that commits to slashing an account and
+front-run it with their own slash transaction before the original slash executes. The commit-reveal scheme prevents this
+by:
+
+1. Separating the decision to slash (commit) from the execution (reveal)
+2. Enforcing a time delay between commitment and revelation
+3. Ensuring all slashes on an account are processed in order
+
+### Step 1: Commit Phase
+
+When a slasher initiates a slash operation, they call `slashCommit()`:
+
+1. **Hash Generation**: The slasher creates a keccak256 hash of the private key and reward recipient address.
+2. **Commitment Recording**: The contract stores this hash along with the timestamp in a mapping.
+3. **Queue Management**: If this is not the first slash commit on this account, the reveal start time is scheduled for
+   after the current reveal window closes.
+4. **Event Emission**: The system records the commitment for later verification.
+
+The hash ensures the slasher commits to specific values without revealing them until the reveal phase.
+
+### Step 2: Reveal Phase
+
+After the reveal window has passed, the slasher calls `slashReveal()` to execute the slash:
+
+1. **Data Provision**: The slasher provides the actual private key and reward recipient address.
+2. **Hash Verification**: The contract verifies that the hash of these values matches the previously committed hash.
+3. **Window Validation**: The contract checks that the current time is within or after the reveal window for this
+   commit.
+4. **Slash Execution**: If all validations pass, the `slash()` function is called to execute the actual penalty.
+
+### Reveal Window Timing
+
+The reveal window ensures orderly processing:
+
+- **First Commit**: The reveal window starts immediately.
+- **Subsequent Commits**: Each new commit on the same account is scheduled to start after the previous window ends.
+- **Window Duration**: Configurable by the admin (default 1 hour).
+- **Window Range**: Minimum 1 second, maximum 1 day.
+
+This queuing mechanism ensures that even if multiple slashes are committed for the same account, they are processed
+sequentially in the order they were committed.
+
+### Direct Slash (Emergency)
+
+The contract also supports a direct `slash()` function for emergency situations or when the slasher wishes to bypass the
+commit-reveal mechanism. However, this requires explicit authorization and is typically reserved for privileged roles.
+
+### Access Control
+
+Only accounts with the `SLASHER_ROLE` or the admin can execute slashing operations.
+
+### Configurable Parameters
+
+- **Reveal Window Duration**: The amount of time that must pass between a commit and a reveal.
+
+## Data Structures and State
```

Comment:
Let's not put this here. If I want to learn about data structures used, I read the code.